packages/SwingSet/src/kernel/kernelSyscall.js

```diff
@@ -93,137 +75,33 @@ export function makeKernelSyscallHandler(tools) {
-   *   the iteration, or '' if this is the first cycle
-   * @param {string} lowerBound  The lower bound of the iteration range
-   * @param {string} [upperBound]  The uppper bound of the iteration range.  If
-   *   omitted, this defaults to a string equivalent to the `lowerBound` string
-   *   with its rightmost character replaced by the lexically next character.
-   *   For example, if `lowerBound` is 'hello', then `upperBound` would default
-   *   to 'hellp')
-   *
-   * @returns {['ok', [string, string]|[undefined, undefined]]} A pair of a
-   *   status code and a result value.  In the case of this operation, the
-   *   status code is always 'ok'.  The result value is a pair of the key and
-   *   value of the first vatstore entry whose key is lexically greater than
-   *   both `priorKey` and `lowerBound`.  If there are no such entries or if the
-   *   first such entry has a key that is lexically greater than or equal to
-   *   `upperBound`, then result value is a pair of undefineds instead,
-   *   signalling the end of iteration.
-   *
-   * Usage notes:
-   *
-   * Iteration is accomplished by repeatedly calling `vatstoreGetAfter` in a
-   * loop, each time pass the key that was returned in the previous iteration,
-   * until undefined is returned.  While the initial value for `priorKey` is
-   * conventionally the empty string, in fact any value that is less than
-   * `lowerBound` may be used to the same effect.
-   *
-   * Keys in the vatstore are arbitrary strings, but subsets of the keyspace are
-   * often organized hierarchically.  This iteration API allows simple iteration
-   * over an explicit key range or iteration over the set of keys with a given
-   * prefix, depending on how you use it:
-   *
-   * Explicitly providing both a lower and an upper bound will enable iteration
-   * over the key range `lowerBound` <= key < `upperBound`
+   * @param {string} priorKey A key before the desired key
    *
-   * Providing only the lower bound while letting the upper bound default will
-   * iterate over all keys that have `lowerBound` as a prefix.
-   *
-   * For example, if the stored keys are:
-   * bar
-   * baz
-   * foocount
-   * foopriority
-   * joober
-   * plugh.3
-   * plugh.47
-   * plugh.8
-   * zot
-   *
-   * Then the bounds    would iterate over the key sequence
-   * ----------------   -----------------------------------
-   * 'bar', 'goomba'    bar, baz, foocount, foopriority
-   * 'bar', 'joober'    bar, baz, foocount, foopriority
-   * 'foo'              foocount, foopriority
-   * 'plugh.'           plugh.3, plugh.47, plugh.8
-   * 'baz', 'plugh'     baz, foocount, foopriority, joober
-   * 'bar', '~'         bar, baz, foocount, foopriority, joober, plugh.3, plugh.47, plugh.8, zot
+   * @returns {['ok', string|null]} A pair of a status code and
+   *   a result value. In the case of this operation, the status code
+   *   is always 'ok'. The result value is either a key or undefined.
```

Comment:
Good catch, I need to trace through where we use each. I remember making a bunch of "no results" results use `null` because the transcript recording mechanism didn't do a good job remembering `undefined` (like maybe `['ok', undefined]` got serialized as `['ok]]`, or transcript replay comparison was seeing false mismatches). But that value is copied through like five types on the way through from kernel to vat to transcript, etc.

I'll track down the full path and figure out the right value to use for this.

Reply:
Ok, so I think the issue was when we write a transcript entry to the `streamStore`, we serialize it with `JSON.stringify` (specifically with [`lib/djson.js`](https://github.com/Agoric/agoric-sdk/blob/aa8e813c26ca8fd93118ed17c0c147acb51747de/packages/SwingSet/src/lib/djson.js)), and `undefined` does not survive a round-trip through JSON (it always comes back as `null`). Transcripts include syscall results, and we need them to match, so we can't afford to have the first (live) result be `['ok', undefined]`, and the second (replayed) result be `['ok', null]`.

So the workaround is to pre-emptively convert syscall results from `undefined` to `null` at some point in the process. We could do any of:
* change swing-store to return `null` instead of `undefined`
* change `kernelSyscall.js` to convert the swing-store's responses to `null`
* change `vatTranslator.js` to convert the `kernelSyscall.js` responses to `null`

I think it feels pretty natural to have the swing-store API use `undefined`, so I'd rather not change it.

It looks like we're using both of the second two options: `kernelSyscall.js` converts responses to `null` (and anyways most of the responses are `['ok', null]` anyways, since most syscall don't have return values, just an `ok`), and also `vatTranslator.js` converts responses to `null` just in case.

I'll change the comment to say "key or null", and I'll make sure each syscall *and* each result translator does the conversion, and I'll see if I can tighten the types to enlist `tsc`'s help in rejecting `undefined` here.

And it reminds me that we should check the syscall *inputs* to make sure they only have nulls, not undefineds.